core/camel-support/src/main/java/org/apache/camel/support/IntrospectionSupport.java

```diff
@@ -646,11 +649,43 @@ public static boolean setProperty(CamelContext context, TypeConverter typeConver
                     value = CamelContextHelper.lookup(context, s);
                 }
                 if (isNotEmpty(lookupKey)) {
-                    int idx = Integer.valueOf(lookupKey);
-                    list.add(idx, value);
+                    int idx = Integer.parseInt(lookupKey);
+                    if (idx < list.size()) {
+                        list.set(idx, value);
+                    } else if (idx == list.size()) {
+                        list.add(value);
+                    } else {
+                        if (list instanceof ArrayList) {
```

Comment:
Maybe add a code comment why you do this, eg to expand the array when there is gaps in the index